### ISSUE TITLE
Tomcat cannot load an SSL KeyStore from a `file:` URL when using an exploded module with JDK 11 on Windows

### ISSUE THREAD

**Issue by Captain-P-Goldfish** [spring-projects/spring-boot#17841]
Configure a spring-boot (I am currently using 2.1.6-RELEASE) project with a tomcat-connector and configure a ssl-keystore within the application.properties that uses either a file-url or an absolute windows-path. What we get is this:

```
java.nio.file.InvalidPathException: Illegal char <:> at index 4: file:\D:\Projekte\intern\ausweis-ident\my_configuration\ausweis-ident.jks
	at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
	at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
	at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:229)
	at java.base/jdk.internal.module.Resources.toSafeFilePath(Resources.java:143)
	at java.base/jdk.internal.module.Resources.toFilePath(Resources.java:97)
	at java.base/jdk.internal.module.ModuleReferences$ExplodedModuleReader.find(ModuleReferences.java:382)
	at java.base/jdk.internal.loader.BuiltinClassLoader$2.run(BuiltinClassLoader.java:409)
	at java.base/jdk.internal.loader.BuiltinClassLoader$2.run(BuiltinClassLoader.java:404)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/jdk.internal.loader.BuiltinClassLoader.findMiscResource(BuiltinClassLoader.java:403)
	at java.base/jdk.internal.loader.BuiltinClassLoader.findResource(BuiltinClassLoader.java:290)
	at java.base/java.lang.ClassLoader.getResource(ClassLoader.java:1400)
	at java.base/java.lang.ClassLoader.getResourceAsStream(ClassLoader.java:1736)
	at tomcat.embed.core@9.0.22/org.apache.catalina.startup.CatalinaBaseConfigurationSource.getResource(CatalinaBaseConfigurationSource.java:91)
	at tomcat.embed.core@9.0.22/org.apache.tomcat.util.net.SSLUtilBase.getStore(SSLUtilBase.java:198)
	at tomcat.embed.core@9.0.22/org.apache.tomcat.util.net.SSLHostConfigCertificate.getCertificateKeystore(SSLHostConfigCertificate.java:206)
	at tomcat.embed.core@9.0.22/org.apache.tomcat.util.net.SSLUtilBase.getKeyManagers(SSLUtilBase.java:280)
	at tomcat.embed.core@9.0.22/org.apache.tomcat.util.net.SSLUtilBase.createSSLContext(SSLUtilBase.java:247)
	at tomcat.embed.core@9.0.22/org.apache.tomcat.util.net.AbstractJsseEndpoint.createSSLContext(AbstractJsseEndpoint.java:97)
	at tomcat.embed.core@9.0.22/org.apache.tomcat.util.net.AbstractJsseEndpoint.initialiseSsl(AbstractJsseEndpoint.java:71)
	at tomcat.embed.core@9.0.22/org.apache.tomcat.util.net.NioEndpoint.bind(NioEndpoint.java:218)
	at tomcat.embed.core@9.0.22/org.apache.tomcat.util.net.AbstractEndpoint.bindWithCleanup(AbstractEndpoint.java:1124)
	at tomcat.embed.core@9.0.22/org.apache.tomcat.util.net.AbstractEndpoint.start(AbstractEndpoint.java:1210)
	at tomcat.embed.core@9.0.22/org.apache.coyote.AbstractProtocol.start(AbstractProtocol.java:585)
	at tomcat.embed.core@9.0.22/org.apache.catalina.connector.Connector.startInternal(Connector.java:1005)
	at tomcat.embed.core@9.0.22/org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at tomcat.embed.core@9.0.22/org.apache.catalina.core.StandardService.addConnector(StandardService.java:227)
	at spring.boot@2.1.7.RELEASE/org.springframework.boot.web.embedded.tomcat.TomcatWebServer.addPreviouslyRemovedConnectors(TomcatWebServer.java:263)
	at spring.boot@2.1.7.RELEASE/org.springframework.boot.web.embedded.tomcat.TomcatWebServer.start(TomcatWebServer.java:195)
	at spring.boot@2.1.7.RELEASE/org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.startWebServer(ServletWebServerApplicationContext.java:297)
	at spring.boot@2.1.7.RELEASE/org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.finishRefresh(ServletWebServerApplicationContext.java:163)
	at spring.context@5.1.6.RELEASE/org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:552)
	at spring.boot@2.1.7.RELEASE/org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:141)
	at spring.boot@2.1.7.RELEASE/org.springframework.boot.SpringApplication.refresh(SpringApplication.java:743)
	at spring.boot@2.1.7.RELEASE/org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:390)
	at spring.boot@2.1.7.RELEASE/org.springframework.boot.SpringApplication.run(SpringApplication.java:312)
	at spring.boot@2.1.7.RELEASE/org.springframework.boot.SpringApplication.run(SpringApplication.java:1214)
	at spring.boot@2.1.7.RELEASE/org.springframework.boot.SpringApplication.run(SpringApplication.java:1203)
	at de.governikus.ausweis.ident/de.governikus.ausweis.ident.AusweisIDent.main(AusweisIDent.java:36)

```

The reason for this can be found within SslConnectorCustomizer within spring-boot jar on line 125.
```

try {
	protocol.setKeystoreFile(ResourceUtils.getURL(ssl.getKeyStore()).toString());
}
```

The path is always resolved to URL and eventually passed by tomcat to the WindowsPathParser that is not able of handling this URL.

This is currently an absolute blocker for us. We need to migrate our project to jdk11 but cannot get past this problem

**Comment by kedar-joshi** [comment]
I not able to reproduce this. Can you please share your `application.properties` or a sample project to reproduce the behaviour ?

**Comment by Captain-P-Goldfish** [comment]
yes I have pushed it to the following repository
https://github.com/Captain-P-Goldfish/spring-boot-path-error
the application.properties can be found in the directory "my_configuration"

**Comment by kedar-joshi** [comment]
I ran the project from the repository with `mvn spring-boot:run` and application started successfully. Here are the steps I took -

1. Copied `application.properties` from `my_configuration` to `src/main/resources`.
2. Copied `my_configuration` directory to `D:/Projekte/intern/ausweis-ident` so as to have correct path.
3. Executed `mvn spring-boot:run` on JDK 11 and below is the output -

```
13 Aug 2019 11:35:25,562 INFO  org.springframework.boot.web.embedded.tomcat.TomcatWebServer (TomcatWebServer.java:202) - Tomcat started on port(s): 8080 (http) with context path ''
13 Aug 2019 11:35:25,581 INFO  de.governikus.ausweis.ident.AusweisIDent (StartupInfoLogger.java:59) - Started AusweisIDent in 2.794 seconds (JVM running for 6.4)
13 Aug 2019 11:35:25,581 INFO  de.governikus.ausweis.ident.AusweisIDent (AusweisIDent.java:47) -

        ,--.                  ,--.            ,--.                                                    ,---.        ,--.,--.
 ,---.,-'  '-.,--,--.,--.--.,-'  '-. ,---.  ,-|  |     ,---. ,--.,--. ,---. ,---. ,---.  ,---.  ,---./  .-',--.,--.|  ||  |,--. ,--.
(  .-''-.  .-' ,-.  ||  .--''-.  .-'| .-. :' .-. |    (  .-' |  ||  || .--'| .--'| .-. :(  .-' (  .-'|  `-,|  ||  ||  ||  | \  '  /
.-'  `) |  | \ '-'  ||  |     |  |  \   --.\ `-' |    .-'  `)'  ''  '\ `--.\ `--.\   --..-'  `).-'  `)  .-''  ''  '|  ||  |  \   '
`----'  `--'  `--`--'`--'     `--'   `----' `---'     `----'  `----'  `---' `---' `----'`----' `----'`--'   `----' `--'`--'.-'  /
                                                                                                                           `---'
```

P.S. I had to hardcode `configurationPort` in `WebSecurityConfiguration` as `@Value("${server.port}")` wasn't working.

**Comment by Captain-P-Goldfish** [comment]
alright... now I am confused why I cannot get it to work...
But thx for your effort.

**Comment by Captain-P-Goldfish** [comment]
the only difference here I see is that I started it from my IDE and not directly with maven

**Comment by Captain-P-Goldfish** [comment]
Tried it now. The problem does really only occur if I start the application with the IDE. The built jar is also working

**Comment by wilkinsona** [comment]
Thanks for the report. I think this may need to be fixed/worked around in Tomcat. `org.apache.catalina.startup.CatalinaBaseConfigurationSource.getResource(String)` tries the different ways of accessing the resource:

1. As a `File`
2. As a resource `InputStream` retrieved from the `ClassLoader`
3. As an `InputStream` created from a `URI`

Boot relies on the third of these but it is not being attempted due to the second failing by throwing an exception rather than returning `null`. I think it could even be argued that this is a JDK bug as I'd expect `null` to be returned (indicating that the resource with the name cannot be found) rather than an exception being thrown due to an implementation detail of how that name is being interpreted.

@markt-asf what's your take on this please?

**Comment by wilkinsona** [comment]
@Captain-P-Goldfish In the meantime, you should be able to work around the problem with a bean similar to the following:

```java
@Bean
WebServerFactoryCustomizer<TomcatServletWebServerFactory> tomcatCustomizer(ServerProperties serverProperties,
        ResourceLoader resourceLoader) {
    return (tomcat) -> tomcat.setSslStoreProvider(new SslStoreProvider() {

        @Override
        public KeyStore getKeyStore() throws Exception {
            KeyStore keyStore = KeyStore.getInstance("JKS");
            try (InputStream storeStream = resourceLoader.getResource("classpath:sample.jks").getInputStream()) {
                keyStore.load(storeStream, serverProperties.getSsl().getKeyStorePassword().toCharArray());
            }
            return keyStore;
        }

        @Override
        public KeyStore getTrustStore() throws Exception {
            return null;
        }

    });
}
```

The above would replace the configuration of the `server.ssl.key-store` in your `application.properties`.

**Comment by markt-asf** [comment]
The sample project that reproduces this has been removed. I'd like to be able to take a closer look at that to understand exactly what is going on first.

**Comment by wilkinsona** [comment]
This [zip file](https://github.com/spring-projects/spring-boot/files/3509887/gh-17841.zip) contains an application that should reproduce the problem if you import it into Eclipse and run its main method. I used Windows 7 and openjdk version "11.0.4" 2019-07-16.

Key to the failure is there being an exploded module on the module path. In the case of the reproduction, it's the application itself that's the exploded module. The exploded module is required as the faulty code path lies beneath `ExplodedModuleReader.find()`.

**Comment by Captain-P-Goldfish** [comment]
oh okay, I guess I do not need to provide the sample project again.
Hope this is getting fixes soon :-)

**Comment by markt-asf** [comment]
Thanks for the test case. A good test case always makes an issue so much easier to debug.

The exception looks legitimate in that what is being passed isn't in the format Java expects for a file URL on Windows.

That first led me to `application.properties`. The value there doesn't look right. I'd expect it to be:
```
server.ssl.key-store: file:///C:/Projekte/intern/ausweis-ident/my_configuration/ausweis-ident.jks
```

I made that change but it didn't make any difference. I then debugged into the Tomcat code that uses that property and I found that, even with the value above, `Http11NioProtocol.setKeystoreFile(String)` is being called with a value of
```
file:/C:/Projekte/intern/ausweis-ident/my_configuration/ausweis-ident.jks
```
If I manually change this in the debugger to:
```
file:///C:/Projekte/intern/ausweis-ident/my_configuration/ausweis-ident.jks
```
then it still fails but with the expected `java.io.FileNotFoundException`.
This looks to have a root cause somewhere around line 125 of SslConnectorCustomizer.class and the use of ResourceUtils.getURL().

**Comment by wilkinsona** [comment]
Thank you, @markt-asf.

The form that you expect for the file URL on Windows doesn't seem to match what the JVM produces. For example, if I add the following line to the repro's main method:

```
System.out.println(DemoApplication.class.getClassLoader().getResource("application.properties"));
```

It produces the following output:

```
file:/C:/Users/Andy%20Wilkinson/Documents/workspace-spring-tool-suite-4-4.3.2.RELEASE/demo/target/classes/application.properties
```

Alternatively, if I create a file and then get its URL:

```
System.out.println(new File("src/main/resources/application.properties").toURI().toURL());
```

The output is similar:

```
file:/C:/Users/Andy%20Wilkinson/Documents/workspace-spring-tool-suite-4-4.3.2.RELEASE/demo/src/main/resources/application.properties
```

This is, essentially, what `ResourceUtils.getURL()` will do for a `classpath:` input and a file path input respectively.

If I update the sample to add a keystore in `src/main/resources` and to use `classpath:example.jks` for `server.ssl.key-store` in `application.properties`, the URL that's passed into Tomcat is the `toString()` of what is returned from `ClassLoader.getResource()` and it fails with the `InvalidPathException`.

**Comment by markt-asf** [comment]
Yep, my mistake. I wonder where I got the `file:///C:/`idea from. It wasn't recently. Anyway, given that Tomcat is already trapping `URISyntaxException` I think we should trap this too. I'll look into a patch.

**Comment by markt-asf** [comment]
Fixed in 9.0.x for 9.0.25 onwards - should be released early September.

**Comment by wilkinsona** [comment]
Thanks very much, @markt-asf. I'll close this one as a Tomcat issue. We'll pick up 9.0.25 in due course.
